### ISSUE TITLE
Bug: [no-deprecated] doesn't report when deprecated API's are used in an object.

### ISSUE THREAD

**Issue by ravichandra480** [typescript-eslint/typescript-eslint#10643]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=PTAEGEEEGUFFQEwC5QDMD2AnUBLAdgC4CmmqAhgMZEDOAsAFD7GmVGgDimOAJgPIAOBHOjzVQAbwahQwAFSzQAAW5F%2BmIhTLFuoAK7U2eIgHdI-HLlHEyO2cCmgyAfhTUCXPAHMA3A4BGLqBuHj4MAL4MDCCgAHLoBKDGWADW%2BJ5BVHhkXOgMFCJuoAAeKJw8AkIFoAC8Eg5kKABEGOiNMmAA6pAASjFNZXyCwqIAdGS4YipqGlpE3Cj6bH6WbkQ24ZH00R0paRlEWTl5BQkAXjXFY95AA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNFjpEZAIb5E3dFETRoHaJHBgAviEVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// CASE 2: for interfaces
interface GridOptions {
  /** @deprecated use newApi instead */
  a?: string;
  b?: string;
}

// Not working scenario
const x: GridOptions = {
  a: "foo" // WARN: "GridOptions.a is deprecated: use b instead
}

// Working scenario
const z = x.a;
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-deprecated": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Linter should report when a deprecated property is used inside an object

### Actual Result

The linter is not reporting when a deprecated property is used inside an object.

### Additional Info

Similar issue was reported [[deprecation] Warns when deprecated APIs are used.](https://github.com/typescript-eslint/typescript-eslint/issues/1223), and this issue was closed by refering to the [eslint-plugin-deprecation](https://github.com/gund/eslint-plugin-deprecation).

**Comment by ronami** [comment]
Looking into this - I'm not sure if this is something the rule should report.

To my understanding, the rule should report cases where TypeScript also shows an API as deprecated. In the provided example, TypeScript only shows line 14 as deprecated (which the rule reports). However, if the rule were to report line 10 (which this issue suggests is the rule report), it would report a case that TypeScript doesn't show as deprecated.

Additionally, line 10 (which this issue suggests the rule report) doesn't actually "use" the deprecated API. I think it's valid (at least in some cases) to assign a value to an interface, even if it marked it as deprecated.

**Comment by JoshuaKGoldberg** [comment]
🤔 maybe this is a bug to be filed on TypeScript? I would have thought the `a` property on line 10 would be marked as deprecated.

**Comment by bradzacher** [comment]
If you hover the expected property then TS shows you the JSDoc including the `@deprecated` tag.*So ts can see the information and so we could too and thus we could act on it. Which was why I tagged it as accepting PRs.

This would likely be a matter of us not using the contextual type of the object and instead we're just using the standard type?

**Comment by ronami** [comment]
> If you hover the expected property then TS shows you the JSDoc including the `@deprecated` tag.*So ts can see the information and so we could too and thus we could act on it. Which was why I tagged it as accepting PRs.
> 
> This would likely be a matter of us not using the contextual type of the object and instead we're just using the standard type?

That's correct. I wasn't sure about this since TypeScript doesn't indicate it as deprecated usage (no strikethrough and no deprecated report from TypeScript).

It looks like there's a TypeScript issue for this: https://github.com/microsoft/TypeScript/issues/60442.
